src/qtism/data/AssessmentTest.php

```diff
@@ -430,17 +433,859 @@ public function isExclusivelyLinear()
+                            }
+
+                            if ($path[$identifier] == $key_current_item) {
+                                $delete_keys = true;
+                            }
+                        }
+
+                        if (!in_array($new_path, $paths)) {
+                            $new_paths[] = $new_path;
+                        }
+                    }
+
+                }
+            } else {
+                if ($prev_item == null) { // Branching starts at the beginning of the test
+
+                    foreach ($paths as $path) {
+
+                        // get the index of the current item and of the target item
+
+                        $key_target_item = null;
+
+                        $pathkeys = $path->getKeys();
+
+                        foreach ($pathkeys as $identifier) {
+
+                            if ($target_item->getIdentifier() == $identifier) {
+                                $key_target_item = $target_item->getIdentifier();
+                            }
+                        }
+                        
+                        if ($key_target_item != null) {
+
+                            $new_path = new AssessmentItemRefCollection($path->getArrayCopy());
+
+                            // Delete from new path everything between $key_current_item and $key_target_item
+
+                            foreach ($pathkeys as $identifier) {
+
+                                if ($identifier == $key_target_item) {
+                                    break;
+                                }
+
+                                unset($new_path[$identifier]);
+                            }
+
+                            if (!in_array($new_path, $paths)) {
+                                $new_paths[] = $new_path;
+                            }
+                        }
+                    }
+
+                } else { // Normal case
+
+                    foreach ($paths as $path) {
+
+                        // get the index of the current item and of the target item
+
+                        $key_current_item = null;
+                        $key_target_item = null;
+
+                        $pathkeys = $path->getKeys();
+
+                        foreach ($pathkeys as $identifier) {
+
+                            if ($prev_item->getIdentifier() == $identifier) {
+                                $key_current_item = $prev_item->getIdentifier();
+                            }
+
+                            if ($target_item->getIdentifier() == $identifier) {
+                                $key_target_item = $target_item->getIdentifier();
+                            }
+                        }
+
+                        if (($key_current_item != null) and ($key_target_item != null)) {
+
+                            if ($key_current_item == $key_target_item) {
+                                throw new BranchRuleTargetException("Recursive branching is not allowed.");
```

Comment:
Example of Exception code usage:

```php
throw new BranchRuleTargetException(
    "Recursive branching is not allowed.",
   BranchRuleTargetException::RECURSIVE_BRANCHING
);
```
It would be awesome if another constructor parameter of BranchRuleTargetException was the AssessmentItemRef instance that was currently scanned. This would give the opportunity to the client code to highlight where the problem occurs! Thanks !